src/main/java/winstone/Launcher.java

```diff
@@ -229,9 +231,19 @@ private void writePortToFileIfNeeded() throws IOException {
                     Path portFile = Paths.get(portFileName);
                     Path portDir = portFile.getParent();
                     Files.createDirectories(portDir);
-                    try (BufferedWriter writer = Files.newBufferedWriter(portFile, StandardCharsets.UTF_8)) {
+                    Path tmpPath = Files.createTempFile(portDir, "temp", "");
+                    try (BufferedWriter writer = Files.newBufferedWriter(tmpPath, StandardCharsets.UTF_8)) {
                         writer.write(Integer.toString(port));
                     }
+
+                    try {
+                        Files.move(tmpPath, portFile, StandardCopyOption.ATOMIC_MOVE);
```

Comment:
In the first attempt, you are not using `StandardCopyOption.REPLACE_EXISTING`, but in the second attempt you are. I see no reason for this inconsistency: the semantics of atomicity are unrelated to the semantics of replacing an existing destination.

I would suggest making both use the same semantics: either `StandardCopyOption.REPLACE_EXISTING` in both attempts or in neither. Since the original code before your PR used `Files.newBufferedWriter`, which uses `StandardOpenOption.TRUNCATE_EXISTING`, the least risky change would be to retain the same semantics. So I suggest using `StandardCopyOption.REPLACE_EXISTING` in both attempts, which means adding it here:

```suggestion
                        Files.move(tmpPath, portFile, StandardCopyOption.REPLACE_EXISTING, StandardCopyOption.ATOMIC_MOVE);
```

Reply:
> I would suggest making both use the same semantics: either StandardCopyOption.REPLACE_EXISTING in both attempts or in neither

as per the [javadoc ](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Files.html#move(java.nio.file.Path,java.nio.file.Path,java.nio.file.CopyOption...)) it would be incorrect to use both `REPLACE_EXISTING` and `ATOMIC_MOVE`, and lead to potentially bad assumptions when reading the code that existing files would be replaced (and that flag honoured) which may or may not be the case depending on your filesystem and operating system combination.

if the file is not expected to ever exist then I agree that REPLACE_EXISTING should not be needed for the non atomic case.

Reply:
> as per the [javadoc ](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Files.html#move(java.nio.file.Path,java.nio.file.Path,java.nio.file.CopyOption...))it would be incorrect to use both REPLACE_EXISTING and ATOMIC_MOVE

I disagree with this. I see no evidence that it is incorrect to use both `REPLACE_EXISTING` and `ATOMIC_MOVE`.

Reply:
Oh I think I see what you mean: "all other options are ignored" when using `REPLACE_EXISTING`. So yes, it is redundant to specify both, and `ATOMIC_MOVE` is sufficient.

Reply:
Yes exactly that, sorry I was not clear.